File: 2024/day20/python/solution.py
#!/usr/bin/env python3
import sys
from collections import deque

def parse_grid(input_text):
    grid = []
    start = end = None
    for r, line in enumerate(input_text.strip().split('\n')):
        row = list(line)
        grid.append(row)
        for c, ch in enumerate(row):
            if ch == 'S':
                start = (r, c)
            elif ch == 'E':
                end = (r, c)
    return grid, start, end

def trace_path(grid, start, end):
    """Trace the single path from start to end, returning distance from start for each cell."""
    rows, cols = len(grid), len(grid[0])
    dist = {start: 0}
    queue = deque([start])

    while queue:
        r, c = queue.popleft()
        if (r, c) == end:
            break
        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] != '#':
                if (nr, nc) not in dist:
                    dist[(nr, nc)] = dist[(r, c)] + 1
                    queue.append((nr, nc))

    return dist

def count_cheats(dist, max_cheat_time, min_savings):
    """Count cheats that save at least min_savings picoseconds."""
    count = 0
    track_positions = list(dist.keys())

    for r1, c1 in track_positions:
        d1 = dist[(r1, c1)]
        for r2, c2 in track_positions:
            # Manhattan distance is the cheat cost
            cheat_cost = abs(r2 - r1) + abs(c2 - c1)
            if cheat_cost <= max_cheat_time:
                d2 = dist[(r2, c2)]
                savings = d2 - d1 - cheat_cost
                if savings >= min_savings:
                    count += 1

    return count

def part1(grid, start, end):
    dist = trace_path(grid, start, end)
    return count_cheats(dist, 2, 100)

def part2(grid, start, end):
    dist = trace_path(grid, start, end)
    return count_cheats(dist, 20, 100)

def main():
    with open('../input.txt') as f:
        input_text = f.read()

    grid, start, end = parse_grid(input_text)

    print('Part 1:', part1(grid, start, end))
    print('Part 2:', part2(grid, start, end))

if __name__ == '__main__':
    main()

File: 2024/day20/ruby/solution.rb
#!/usr/bin/env ruby

def parse_grid(input_text)
  grid = []
  start_pos = nil
  end_pos = nil

  input_text.strip.split("\n").each_with_index do |line, r|
    row = line.chars
    grid << row
    row.each_with_index do |ch, c|
      if ch == 'S'
        start_pos = [r, c]
      elsif ch == 'E'
        end_pos = [r, c]
      end
    end
  end

  [grid, start_pos, end_pos]
end

def trace_path(grid, start_pos, end_pos)
  # Trace the single path from start to end, returning distance from start for each cell
  rows = grid.length
  cols = grid[0].length
  dist = { start_pos => 0 }
  queue = [start_pos]

  while !queue.empty?
    r, c = queue.shift
    break if [r, c] == end_pos

    [[-1, 0], [1, 0], [0, -1], [0, 1]].each do |dr, dc|
      nr, nc = r + dr, c + dc
      if nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] != '#'
        unless dist.key?([nr, nc])
          dist[[nr, nc]] = dist[[r, c]] + 1
          queue << [nr, nc]
        end
      end
    end
  end

  dist
end

def count_cheats(dist, max_cheat_time, min_savings)
  # Count cheats that save at least min_savings picoseconds
  count = 0
  track_positions = dist.keys

  track_positions.each do |pos1|
    r1, c1 = pos1
    d1 = dist[pos1]

    track_positions.each do |pos2|
      r2, c2 = pos2
      # Manhattan distance is the cheat cost
      cheat_cost = (r2 - r1).abs + (c2 - c1).abs

      if cheat_cost <= max_cheat_time
        d2 = dist[pos2]
        savings = d2 - d1 - cheat_cost
        count += 1 if savings >= min_savings
      end
    end
  end

  count
end

def part1(grid, start_pos, end_pos)
  dist = trace_path(grid, start_pos, end_pos)
  count_cheats(dist, 2, 100)
end

def part2(grid, start_pos, end_pos)
  dist = trace_path(grid, start_pos, end_pos)
  count_cheats(dist, 20, 100)
end

def main
  input_text = File.read('../input.txt')
  grid, start_pos, end_pos = parse_grid(input_text)

  puts "Part 1: #{part1(grid, start_pos, end_pos)}"
  puts "Part 2: #{part2(grid, start_pos, end_pos)}"
end

main
